pkg/loadbalancer/reconciler/termination.go

```diff
@@ -0,0 +1,252 @@
+	// destroying sockets. This allows the tests to use "registerSocketTermination"
+	// directly and provide custom implementations of these.
+	cell.ProvidePrivate(
+		func() netnsOps {
+			return netnsOps{
+				current: netns.Current,
+				do:      (*netns.NetNS).Do,
+				all:     netns.All,
+			}
+		},
+		func(log *slog.Logger) sockets.SocketDestroyer {
+			return socketDestroyer{log}
+		},
+	),
+
+	cell.Invoke(registerSocketTermination),
+)
+
+type socketTerminationParams struct {
+	cell.In
+
+	JobGroup  job.Group
+	DB        *statedb.DB
+	Backends  statedb.Table[*lb.Backend]
+	Log       *slog.Logger
+	Config    lb.Config
+	ExtConfig lb.ExternalConfig
+	LBMaps    maps.LBMaps
+
+	SocketDestroyer sockets.SocketDestroyer
+	NetNSOps        netnsOps
+
+	// TestSyncChan is used by the tests to synchronize with the job so it
+	// knows when to delete the backend.
+	TestSyncChan testSyncChan   `optional:"true"`
+	TestConfig   *lb.TestConfig `optional:"true"`
+}
+
+type testSyncChan chan struct{}
+
+// netnsOps captures the operations performed with network namespaces.
+// This allows the tests to inject their own implementation.
+type netnsOps struct {
+	current func() (*netns.NetNS, error)
+	do      func(*netns.NetNS, func() error) error
+	all     func() (iter.Seq2[string, *netns.NetNS], <-chan error)
+}
+
+type socketDestroyer struct {
+	log *slog.Logger
+}
+
+func (sd socketDestroyer) Destroy(filter sockets.SocketFilter) error {
+	return sockets.Destroy(sd.log, filter)
+}
+
+func registerSocketTermination(p socketTerminationParams) error {
+	if !p.Config.EnableExperimentalLB {
+		return nil
+	}
+
+	if p.SocketDestroyer == nil {
+		// To make the load-balancer cell easier to use in tests we don't require that
+		// SocketDestroyer is always provided.
+		if p.TestConfig == nil {
+			return fmt.Errorf("SocketDestroyer not provided and not running in tests")
+		}
+		return nil
+	}
+
+	if !(p.ExtConfig.EnableSocketLB || p.ExtConfig.BPFSocketLBHostnsOnly) {
+		return nil
+	}
+
+	p.JobGroup.Add(
+		job.OneShot(
+			"socket-termination",
+			func(ctx context.Context, h cell.Health) error {
+				return socketTerminationLoop(p, ctx, h)
+			},
+		))
+
+	return nil
+}
+
+func socketTerminationLoop(p socketTerminationParams, ctx context.Context, health cell.Health) error {
+	wtxn := p.DB.WriteTxn(p.Backends)
+	changeIter, err := p.Backends.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	// Process the changes in batches every 50 milliseconds.
+	limiter := rate.NewLimiter(50*time.Millisecond, 1)
+	defer limiter.Stop()
+
+	for {
+		changes, watch := changeIter.Next(p.DB.ReadTxn())
+		for change := range changes {
+			backend := change.Object
+
+			if p.TestSyncChan != nil {
+				close(p.TestSyncChan)
+				p.TestSyncChan = nil
+			}
+
+			// Terminate the sockets connected to backends that have been either
+			// deleted or which are no longer considered viable.
+			if change.Deleted || !backend.IsAlive() {
```

Comment:
Is it possible to get an update for a backend even when there was neither a change in the state nor it was deleted?

Reply:
No it is not. The object in the change must have been either `Insert`ed or `Delete`d. Of course the contents of the object could be the same (even though it must be pointing to a different copy in the heap).